src/systems/filecoin_mining/storage_mining/storage_miner_actor.go

```diff
@@ -600,72 +634,79 @@ func (a *StorageMinerActorCode_I) ProveCommitSector(rt Runtime, info sector.Sect
 		rt.Fatal("sm.ProveCommitSector: sector already exists.")
 	}
 
-	// TODO: check on if ProveCommitSector comes too late after PreCommitSector
+	// check if ProveCommitSector comes too late after PreCommitSector
+	elapsedEpoch := rt.CurrEpoch() - preCommitSector.ReceivedEpoch()
 
-	onChainInfo := &sector.OnChainSealVerifyInfo_I{
-		SealedCID_:                 preCommitInfo.SealedCID(),
-		SealRandomness_:            preCommitInfo.SealRandomness(),
-		InteractiveSealRandomness_: info.InteractiveSealRandomness(),
-		Proof_:                     info.Proof(),
-		DealIDs_:                   preCommitInfo.DealIDs(),
-		SectorNumber_:              preCommitInfo.SectorNumber(),
-	}
-
-	isSealVerificationCorrect := st._isSealVerificationCorrect(rt, onChainInfo)
-	if !isSealVerificationCorrect {
-		rt.Fatal("sm.ProveCommitSector: seal verification failed.")
-	}
+	// if fewer than MAX_PROVE_COMMIT_SECTOR_EPOCH has elapsed
+	if elapsedEpoch <= MAX_PROVE_COMMIT_SECTOR_EPOCH {
```

Comment:
we should check the error case first, handle it, return. otherwise proceed ([go error return convention](https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow))

```go
if elapsedEpoch > MAX_PROVE_COMMIT_SECTOR_EPOCH {
    delete(st.PreCommittedSectors(), preCommitSector.Info().SectorNumber())
    h.Commit(st)
    return // error
}

// ok we're good here.

... rest
```